expression/chunk_executor.go

```diff
@@ -0,0 +1,278 @@
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package expression
+
+import (
+	"strconv"
+
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/sessionctx/stmtctx"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+)
+
+// Vectorizable checks whether a list of expressions can employ vectorized execution.
+func Vectorizable(exprs []Expression) bool {
+	for _, expr := range exprs {
+		if hasUnVectorizableFunc(expr) {
+			return false
+		}
+	}
+	return true
+}
+
+// hasUnVectorizableFunc checks whether an expression contains functions that can not utilize the vectorized execution.
+func hasUnVectorizableFunc(expr Expression) bool {
+	scalaFunc, ok := expr.(*ScalarFunction)
+	if !ok {
+		return false
+	}
+	if scalaFunc.FuncName.L == ast.SetVar {
+		return true
+	}
+	for _, arg := range scalaFunc.GetArgs() {
+		if hasUnVectorizableFunc(arg) {
+			return true
+		}
+	}
+	return false
+}
+
+// UnVectorizedExecute evaluates a list of expressions column by column and append their results to "output" Chunk.
+func VectorizedExecute(ctx context.Context, exprs []Expression, input, output *chunk.Chunk) error {
+	sc := ctx.GetSessionVars().StmtCtx
+	for colID, expr := range exprs {
+		err := evalOneColumn(sc, expr, input, output, colID)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+func evalOneColumn(sc *stmtctx.StatementContext, expr Expression, input, output *chunk.Chunk, colID int) error {
+	switch fieldType, evalType := expr.GetType(), expr.GetType().EvalType(); evalType {
+	case types.ETInt:
+		for rowID, length := 0, input.NumRows(); rowID < length; rowID++ {
```

Comment:
We can switch to assign a function to avoid loop in each case.
```
var execFunc func
switch {
case types.ETInt:
    execFunc = executeToInt
}
for rowID, length := 0, input.NumRows(); rowID < length; rowID++   {
    execFunc()
}

```

Reply:
I think it is already simple enough and there is no need to bring the latency of function addressing.